Simple OPTIONS

ksdenisova opened the pull request:
**Feature**: Simple OPTIONS

  **Scenario**: Finding OPTIONS for an endpoint with only GET
    **Given** I make an OPTIONS request to "/method_options"
    **Then** my response should have status code 200
    **And** my response should have allowed headers of GET, HEAD, and OPTIONS
    **And** my response should have an empty body

  **Scenario**: Finding OPTIONS for an endpoint with multiple verbs
    **Given** I make an OPTIONS request to "/method_options2"
    **Then** my response should have status code 200
    **And** my response should have allowed headers of GET, HEAD, OPTIONS, PUT, and POST
    **And** my response should have an empty body

* Figuring out the OPTIONS is the work of HTTPServer
* For '/method_options' only implement GET method
* For '/method_options2' only implement GET, PUT and POST methods
* PUT is only route is not implemented

[Trello](https://trello.com/c/J1jtdIC9/13-simple-options)

**Acceptance tests**
<img width="1305" alt="image" src="https://user-images.githubusercontent.com/89826596/148082926-996294b3-6b59-4f30-9aab-f829d105a585.png">

**Postman**
<img width="398" alt="image" src="https://user-images.githubusercontent.com/89826596/148416384-830c0ea9-a547-45d9-b74f-997e7f4045f8.png">


**Refactoring:**
- Refactor `Routes` to return `nil` instead throwing an Exception.
- Remove the `MethodNotFound` exception. `Router` generates 404 response (in next story, I'll change it to different response with status code 405).
- Move `HEAD` logic to the function `build_head_response`.
- Update Acceptance test to expect 204 status code instead 200 for `OPTIONS` response.

Aoxian commented on lib/routes.ex:
```diff
@@ -5,7 +5,7 @@ defmodule Routes do
   def handle(%Request{resource: "/echo_body", method: method}) do
     case method do
-      "POST" ->
+      method when method == "POST" or method == "OPTIONS" ->
```
I better understand the question you brought up in today's standup.  

`routes` is essentially a configuration file.  It it a way to declare what routes are available for an implementation that uses the http server framwork you are creating.  It should not be aware of HEAD or OPTIONS, and these should be generated by the http server.  

I would start by refactoring the OPTIONS logic out of this file.  The HEAD logic currently lives in `lib/http_server/router.ex`, so maybe start there.

Aoxian replied:
What you are really looking for in this story, is a way to be aware of what is implemented in this configuration file.  There are a couple of approaches you could take. Since this is a functional language, could you write a function like this:

`is_implemented(routes, method, path)`

So that you could call it like so:

```
assert is_implemented(routes, `/method_options` GET)  == true
assert is_implemented(routes, `/method_options` POST)  == false
assert is_implemented(routes, `/method_options2` POST)  == false
```

Once you have this function available, think about how you can use it to build an OPTIONS response for any configured endpoint.

ksdenisova replied:
I understand that `OPTIONS` should be in the `Routes`. Thank you for the idea, I'm going to refactor it today.

Aoxian replied:
HINT: `Routes.allowed_methods/1` is not configuration, it is also a function that belongs outside of this file, or perhaps not exist at all.

ksdenisova replied:
I removed `allowed_methods` from `Routes`. `Router` checks if it can find a handler for the provided endpoint for all supported HTTP methods, if it is supported, `Router` adds the method to the Set. I use this Set with allowed methods to build a response to the `OPTIONS` request.
I also moved `HEAD` logic to the function `build_head_response`. `Router` builds `HEAD` response only if `GET` method is allowed for the endpoint. It works correctly now. 
I understand that `Router` looks more complicated, but this is the best solution I have found so far.

ksdenisova commented:
I moved `OPTIONS` to the `Router` and added `supported?/2` function. I don't like that I call `find_handler` twice, maybe refactor it later. Also `HEAD` method doesn't work correctly with this implementation. Acceptance tests pass, but I see it in `Postman`. I will continue to work on this.

Aoxian commented on lib/routes.ex:
```diff
@@ -1,54 +1,24 @@
-        raise Exception.MethodNotFoundError
-    end
-  end
+  def handle(%Request{resource: "/echo_body", method: "POST"}), do: Handler.SimplePost
+
+  def handle(%Request{resource: "/simple_get", method: "GET"}), do: Handler.SimpleGet
+
+  def handle(%Request{resource: "/simple_get_with_body", method: "GET"}), do: Handler.SimpleGetWithBody
+
+  def handle(%Request{resource: "/redirect", method: "GET"}), do: Handler.SimpleRedirect
+
+  def handle(%Request{resource: "/head_request", method: "GET"}), do: Handler.HeadRequest
+
+  def handle(%Request{resource: "/method_options", method: "GET"}), do: Handler.SimpleGet
+
+  def handle(%Request{resource: "/method_options2", method: "GET"}), do: Handler.SimpleGet
+
+  def handle(%Request{resource: "/method_options2", method: "POST"}), do: Handler.SimplePost
+
+  def handle(%Request{resource: "/method_options2", method: "PUT"}), do: Handler.SimplePut
```
Now this is really starting to look like a routes file!

ksdenisova replied:
Thank you, it is.

Aoxian commented on lib/http_server/router.ex:
```diff
@@ -2,48 +2,91 @@ defmodule HttpServer.Router do
 
-  def find_handler(routes, %Request{method: _} = request), do: routes.handle(request)
+  def find_allowed_methods(routes \\ Routes, http_methods, resource),
+    do: find_allowed_methods(routes, http_methods, resource, MapSet.new())
+
+  def find_allowed_methods(routes, [method | http_methods], resource, allowed_methods) do
+    allowed_methods =
+      if find_handler(routes, %Request{resource: resource, method: method}) != nil do
+        MapSet.put(allowed_methods, method)
+      else
+        allowed_methods
+      end
 
-  def find_handler(_routes, _request), do: nil
+      find_allowed_methods(routes, http_methods, resource, allowed_methods)
+  end
+
+  def find_allowed_methods(_routes, _methods, _resource, allowed_methods) do
+    allowed_methods
+  end
```
Bravo! Stupendous! Wonderful! Amazing!

ksdenisova replied:
Thank you very much!

Aoxian commented on lib/http_server/router.ex:
```diff
@@ -2,48 +2,91 @@ defmodule HttpServer.Router do
+  alias HttpServer.Response.Headers
   alias HttpServer.Response.Response
 
+  @http_methods ["GET", "POST", "PUT", "CONNECT", "DELETE", "TRACE", "PATCH"]
```
I am curious as to why `CONNECT`, `DELETE`, `TRACE` and `PATCH` made the list here.  Are they actually supported? Could you add a path to `routes.ex` and it would work?

ksdenisova replied:
> Could you add a path to routes.ex and it would work?

What path do you mean?

Aoxian replied:
Hypothetical;
1. create a new `Handler`: `Handler.IansConnect`
2. route to it: `def handle(%Request{resource: "/ians_connect", method: "CONNECT"}), do: Handler.IansConnect`

Would it work?

Aoxian replied:
I think it might

ksdenisova replied:
Yes, it should work. And we don't need to update `@http_methods` to see `CONNECT` method  in `allowed_methods` for this endpoint.

Aoxian replied:
very cool

Aoxian commented on lib/routes.ex:
```diff
@@ -1,54 +1,24 @@
+
+  def handle(%Request{resource: "/method_options", method: "GET"}), do: Handler.SimpleGet
+
+  def handle(%Request{resource: "/method_options2", method: "GET"}), do: Handler.SimpleGet
+
+  def handle(%Request{resource: "/method_options2", method: "POST"}), do: Handler.SimplePost
+
+  def handle(%Request{resource: "/method_options2", method: "PUT"}), do: Handler.SimplePut
```
I wonder could you do something like this?
```elixir
  def handle(%Request{resource: "method_options2"} = request) do
    case request.method do
      "GET" -> Handler.SimpleGet 
      "POST" -> Handler.SimplePost
      "PUT" -> Handler.SimplePut
      _ -> nil
    end
  end
```

Aoxian replied:
Also, and this is not for this PR.  I think we should move away from reusing Handlers in multiple places.  It would be much better to have one Handler per endpoint / method.

```elixir
  def handle(%Request{resource: "method_options2"} = request) do
    case request.method do
      "GET" -> Handler.MethodOptions2.Get
      "POST" -> Handler.MethodOptions2.Post
      "PUT" -> Handler.MethodOptions2.Put
      _ -> nil
    end
  end
```

Or better yet, one `Handler` per endpoint with multiple functions implemented.

```elixir
  def handle(%Request{resource: "method_options2"} = request) do
    case request.method do
      "GET" -> Handler.MethodOptions2.index(request)
      "POST" -> Handler.MethodOptions2.create(request)
      "PUT" -> Handler.MethodOptions2.update(request)
      _ -> nil
    end
  end
```

ksdenisova replied:
> I wonder could you do something like this?
> 
> ```elixir
>   def handle(%Request{resource: "method_options2"} = request) do
>     case request.method do
>       "GET" -> Handler.SimpleGet 
>       "POST" -> Handler.SimplePost
>       "PUT" -> Handler.SimplePut
>       _ -> nil
>     end
>   end
> ```

Yes, I actually did it this way at first, but then changed it to look more similar to the other functions. I can change it back if it is a better way to do this.

Aoxian replied:
I personal preference.  I'd like it because all the methods for one endpoint would be grouped together.  Weight the pros and cons and I'm good with your approach.

ksdenisova replied:
> Also, and this is not for this PR. I think we should move away from reusing Handlers in multiple places. It would be much better to have one Handler per endpoint / method.
> 
> ```elixir
>   def handle(%Request{resource: "method_options2"} = request) do
>     case request.method do
>       "GET" -> Handler.MethodOptions2.Get
>       "POST" -> Handler.MethodOptions2.Post
>       "PUT" -> Handler.MethodOptions2.Put
>       _ -> nil
>     end
>   end
> ```
> 
> Or better yet, one `Handler` per endpoint with multiple functions implemented.
> 
> ```elixir
>   def handle(%Request{resource: "method_options2"} = request) do
>     case request.method do
>       "GET" -> Handler.MethodOptions2.index(request)
>       "POST" -> Handler.MethodOptions2.create(request)
>       "PUT" -> Handler.MethodOptions2.update(request)
>       _ -> nil
>     end
>   end
> ```

I think this is a good idea. We can do this refactoring together in the next pair programming session.

ksdenisova replied:
> I personal preference. I'd like it because all the methods for one endpoint would be grouped together. Weigh the pros and cons and I'm good with your choice.

I've changed it to way we both think about. It looks more descriptive when the all methods are grouped together for one endpoint.
